uncertainties/core.py

```diff
@@ -1756,6 +1756,19 @@ def __bool__(self):
 
     ########################################
 
+    def __hash__(self):
+        """
+        Calculates the hash for any AffineScalarFunc object.
+        The hash is calculated from the nominal_value, and the derivatives.
+        
+        Returns:
+            int: The hash of this object
+        """
+
+        ids = [id(d) for d in self.derivatives.keys()]
+        values_hash = tuple(self.derivatives.values())
+        return hash((self._nominal_value, tuple(ids), values_hash))
```

Comment:
I think we should add a case to `test_hash` like

```python
z_xy = x + y
z_yx = y + x
assert z_xy == z_yx
assert hash(z_xy) == hash(z_yx)
```

That case fails for me with this implementation but the implementation is dependent on the iteration order of `derivatives` which depends on in the insertion order, which can vary depending on the order in which operations are performed. Changing the tuples to `frozenset` or sorting the values before casting to `tuple` could remove the order dependence.

Also, we could define `__hash__` on `LinearCombination` and use that here instead of constructing it from `derivatives` directly. Perhaps there is no use for that though.

Reply:
I think it's worth trying implementations that cover obvious use cases (such as z_xy == z_yx) and using assertions to raise if we've gone past the limit of our assumptions.

Reply:
Hm the problem of a forzenset might be, that the elements of a set must be unique right?
The `ids` are unique, but the values might not.
I will try to create a test for that...

Reply:
This is the example:

```py
a = ufloat(1.23, 2.34)
b = ufloat(1.23, 2.34)
c = ufloat(1.23, 2.34)

x = AffineScalarFunc(1, LinearCombination({a:1, b:2, c:1}))
y = AffineScalarFunc(1, LinearCombination({a:1, b:2, c:2}))
assert x != y                   # Valid
assert hash(x) != hash(y) # Both hashes are identical :(
```

Reply:
I am using sorted tuples now, to avoid this problem.